runtime/binding-kafka/src/main/java/io/aklivity/zilla/runtime/binding/kafka/internal/stream/KafkaClientSaslHandshaker.java

```
@@ -430,8 +568,98 @@ protected final int decodeSaslAuthenticate(
+    }
+
+    protected final int decodeSaslScramAuthenticate(
+            KafkaSaslClient client,
+            long traceId,
+            long authorization,
+            long budgetId,
+            int reserved,
+            DirectBuffer buffer,
+            int offset,
+            int progress,
+            int limit)
+    {
+        final int length = limit - progress;
+
+        if (length != 0)
+        {
+            final SaslAuthenticateResponseFW authenticateResponse = saslAuthenticateResponseRO.tryWrap(buffer, progress, limit);
+
+            if (authenticateResponse != null)
+            {
+                final int errorCode = authenticateResponse.errorCode();
+
+                progress = authenticateResponse.limit();
+
+                client.decodableResponseBytes -= authenticateResponse.sizeof();
+                assert client.decodableResponseBytes >= 0;
+
+                try
+                {
+                    byte[] serverKey = formatter.serverKey(client.saltedPassword);
+                    byte[] serverSignature = formatter.hmac(serverKey, client.authMessage);
+                    DirectBuffer serverFinalResponse = authenticateResponse.authBytes().value();
+                    String serverFinalMessage = serverFinalResponse.getStringWithoutLengthUtf8(0,
+                            serverFinalResponse.capacity());
+
+                    Matcher serverFinalResponseMatcher = SERVER_FINAL_MESSAGE.matcher(serverFinalMessage);
+                    while (serverFinalResponseMatcher.find())
+                    {
+                        serverFinalMessage = serverFinalResponseMatcher.group(1);
+                    }
+                    if (!Arrays.equals(Base64.getDecoder().decode(serverFinalMessage),
+                            serverSignature))
+                    {
+                        client.onDecodeSaslResponse(traceId);
+                        client.onDecodeSaslAuthenticateResponse(traceId, authorization, 58);
```

Avoid hardcoded values like `58` and instead create a static constant for `ERROR_SASL_AUTHENTICATION_FAILED`.

Please make sure we have a negative test in the scripts that sends an invalid verifier value from the server to confirm we behave correctly.